Enable TSDB downsampling ILM configuration

Dosant opened the pull request:
## Summary

Close https://github.com/elastic/kibana/issues/130437

Enables TSDB downsampling action in ILM configuration. 
Actions can be set in the hot, warm, cold phases. 
 - It is hidden in warm and cold phases when searchable snapshots are enabled in hot phase (just like shrink, readonly, force merge actions). 

![Screenshot 2022-08-31 at 17 00 43](https://user-images.githubusercontent.com/7784120/187711538-e67999ff-5e25-4680-8b86-5ddcc64cca23.png)


Users need to specify an interval: 

![Screenshot 2022-08-31 at 17 00 48](https://user-images.githubusercontent.com/7784120/187711593-9540df96-a8d4-43cb-8621-4f50121ff414.png)

Intervals have the following validation: 
-  they must be greater than any previous interval and a multiple of that interval.
   - For example, on the warm phase, the user includes a 3-hour interval; if they choose to add a rollup action on the cold phase, the interval must be a multiple of 3h. It cannot be 5h, for example.
   - Another example: a user can have a rollup on the hot phase of 2d and then no rollup action on the warm phase, and then a rollup action on the cold phase, but it must be a multiple of 2d, for example, it cannot be 1d nor can it be 3d.

![Screenshot 2022-08-31 at 17 07 36](https://user-images.githubusercontent.com/7784120/187712990-03378b93-44af-409f-be1b-cb07f4cbd98b.png)


### How to test

(What I tested)

1. Create an ILM policy with Hot phase only. Enabled rollover, disable "use recommended settings" and set "maximum documents to 1 
2. Enable new "Downsample" action, set interval to 1 hour.
3. Save policy 
4. Then create a data stream template using Dev tools: https://github.com/csoulios/downsampling-demo/blob/main/downsample-ilm.es#L26-L169. Currently, it isn't possible to create this template using UI because of https://github.com/elastic/kibana/issues/139922.  Make sure you are attaching your ilm policy to the template. You can also do it through UI
7. Create data stream by injesting data using Dev tools: https://github.com/csoulios/downsampling-demo/blob/1e66a13bccc7d70aae7763bcc9c4499ec8b3fd02/downsample-ilm.es#L174-L194 . You will have to change timestamp in the data to the current hour 
8. Inspect indices and data streams. You should see them changed during the next hour. Inspect data of the downsampled stream in Discover


## Note

- Rollup of rollups is not yet implemented on es side. We should test everything end to end when it is ready. We wouldn't need to change ILM screen when it ready so we can merge without waiting for that.
- Ideally we should have a es docs link about ILM and downsampling to point to

### Release Notes

Enables time series downsampling action in ILM configuration

Dosant commented:
@elasticmachine merge upstream

Dosant commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

elasticmachine commented:
Pinging @elastic/kibana-app-services (Team:AppServicesUx)

jloleysens commented on x-pack/plugins/index_lifecycle_management/common/types/policies.ts:
```diff
@@ -110,6 +113,8 @@ export interface SerializedWarmPhase extends SerializedPhase {
     shrink?: ShrinkAction;
     forcemerge?: ForcemergeAction;
     readonly?: {};
+    // TODO: rename the key to "downsample" when renamed in es
```
Do we still need these TODOs?

Dosant replied:
Thanks, removed!

jloleysens commented on x-pack/plugins/index_lifecycle_management/public/application/sections/edit_policy/form/validations.ts:
```diff
@@ -272,3 +273,103 @@ export const minAgeGreaterThanPreviousPhase =
+            intervalValues.hot.milliseconds
+          )
+        ) {
+          return {
+            message: i18n.translate(
+              'xpack.indexLifecycleMgmt.editPolicy.downsamplePreviousIntervalWarmPhaseError',
+              {
+                defaultMessage:
+                  'Must be greater than and a multiple of the hot phase value ({value})',
+                values: {
+                  value: intervalValues.hot.esFormat,
+                },
+              }
+            ),
+          };
```
Do we want to provide a pre-filled number for downsampling here? One thing that is tricky about this form is how big it is and I feel like scrolling between phases you can easily get distracted by all the shiny things.

One alternative I thought of was providing a list of multiples in the error message like: (ex. 10 days, 20 days...)

Might not be worth the effort.

Dosant replied:
> Do we want to provide a pre-filled number for downsampling here?

As I understand, this isn't a low effort because this wouldn't be just a static value and have to depend on the previous step: 
e.g., if someone configured a 1-day interval in a hot phase and then enabled a warm phase, then we should prefill with e.g., 2 days. As I understand, such support for "dynamic" initial values needs to be implemented first

Also really not sure what could be good defaults here. 

> One alternative I thought of was providing a list of multiples in the error message like: (ex. 10 days, 20 days...)

Could be!

ghudgins replied:
lets worry about optimizations and better defaults in subsequent PRs. hard to say what the rule should be without spending some time with it / hearing from users

jloleysens reviewed (approved):
Great work @Dosant !

I had two observations I wanted to get y'alls thoughts on (both non-blockers):

First: @yuliacech (and @Dosant) as I was interacting with this beast of a form I thought one thing that could be _really_ useful is to add the ability to filter out actions - maybe like a global fuzzy matcher so that if I type "down" then all the advanced options will auto expand and show only matching actions. This increases the chances that you'll be able to see the continuity between all relevant config more easily. (I'm thinking of this as a future enhancement of course). Similar to the UX you get when searching for settings in VS Code:

<img width="715" alt="Screenshot 2022-09-12 at 10 48 13" src="https://user-images.githubusercontent.com/8155004/189611727-64c65753-12f0-4b30-aaf3-95108db4c311.png">

Second: When I enable the downsampling action for the first time the field starts in an errored state.

<img width="1189" alt="Screenshot 2022-09-12 at 10 25 16" src="https://user-images.githubusercontent.com/8155004/189610121-aa9c8970-1e35-4e44-b026-339b8d12695d.png">

I'm guessing this is because the form is not in a pristine state anymore (you need to enable the action). Would it be possible to have a default value here?

Otherwise, this PR LGTM tested locally.

Dosant commented:
@elasticmachine merge upstream

Dosant commented:
@jloleysens, 

> Second: When I enable the downsampling action for the first time the field starts in an errored state.
> I'm guessing this is because the form is not in a pristine state anymore (you need to enable the action). Would it be possible to have a default value here?

This was because I validated the interval for the current and next phases when downsample action was toggled. I fixed this by validating only the next phases when it is toggled. 
just FYI: We need this for the following scenario:

1. Have hot downsample action enabled and set 1d 
2. Have warm downsample action enabled and set 1d  -> see the error
3. Disable downsample on hot action  (need to trigger validation of warm, so that error resets) 


note: there is the following edge case: 

1. Have the warm downsample action enabled and set 1d 
2. Have the cold downsample action enabled and set 1d  -> see the error
3. Disable the warm step completely -> error on cold downsample action is still there until triggering change or saving. 

To fix this, we need to validate downsample actions when steps are toggled. But I thought this is an overkill for this edge case.

yuliacech commented:
Hey @jloleysens, thanks a lot for this suggestion, I think that is a great idea for the ILM form, it's becoming scarier with each PR 😆  Do you mind opening an issue for that so we don't forget? 

Thanks a lot for working on this feature @Dosant! I had a quick look and the validation worked correctly for me. I agree with JL that the interval input should not be marked as error as soon as the user toggles the Downsampling action. I think there is some similar logic in the MinAge field. The field is required if the phase is enabled but it's nor marked as error straight away. Also there is similar validation logic in this field, because the MinAge can't be smaller than in the previous phase. Also is the Downsampling action also available in the Frozen phase? 

A suggestion I have for a follow up PR: since downsampling turns the index read-only automatically, I think we could hide the ReadOnly action in the phase if there is a downsampling action enabled in it or in any of the previous phases. 

Could you please add the Downsampling action to this text in the Hot phase?
 
<img width="513" alt="Screenshot 2022-09-12 at 14 12 13" src="https://user-images.githubusercontent.com/6585477/189652978-e318d946-c171-4035-808b-297e23e58b2c.png">

Dosant commented:
@yuliacech, 
>I agree with JL that the interval input should not be marked as error as soon as the user toggles the Downsampling action. I think there is some similar logic in the MinAge field. The field is required if the phase is enabled but it's nor marked as error straight away. 

this should be fixed 👍 https://github.com/elastic/kibana/pull/138748#issuecomment-1243637515

> Also is the Downsampling action also available in the Frozen phase?

No! only hot,warm,cold

> A suggestion I have for a follow-up PR: since downsampling turns the index read-only automatically, I think we could hide the ReadOnly action in the phase if there is a downsampling action enabled in it or in any of the previous phases.


Good suggestion. I will create an issue and will try in a separate pr

> Could you please add the Downsampling action to this text in the Hot phase?


👍

alisonelizabeth commented on x-pack/plugins/index_lifecycle_management/__jest__/client_integration/edit_policy/form_validation/downsample_interval.test.ts:
```diff
@@ -0,0 +1,130 @@
+
+    // disable warm phase;
+    await actions.togglePhase('warm');
+    // TODO: there is a bug that disabling a phase doesn't trigger downsample validation in other phases,
```
Is there a separate issue open to address this? Will the validation work as expected on save?

Dosant replied:
> Will the validation work as expected on saving?

Yes! the form state stays in the error state until the validation is triggered again.
More details on the case here: https://github.com/elastic/kibana/pull/138748#issuecomment-1243637515

> note: there is the following edge case:
> Have the warm downsample action enabled and set 1d
> Have the cold downsample action enabled and set 1d -> see the error
> Disable the warm step completely -> error on cold downsample action is still there > until triggering change or saving.
> To fix this, we need to validate downsample actions when steps are toggled. But I thought this is an overkill for this edge case.


> Is there a separate issue open to address this?

I wanted to leave a TODO comment to point this out as a known bug/edge case but didn't think this was worth an issue

Dosant replied:
I'll see if I can fix this with low effort: https://github.com/elastic/kibana/pull/140628

alisonelizabeth commented on x-pack/plugins/index_lifecycle_management/__jest__/client_integration/edit_policy/form_validation/downsample_interval.test.ts:
```diff
@@ -0,0 +1,130 @@
+      ['Must be greater than and a multiple of the hot phase value (60m)'],
+      'cold'
+    );
+  }, /* increase a timeout of this test as it could fail on ci*/ 12000);
```
Out of curiosity - is there any more context you can provide here? Was the test failing without the increased timeout?

Dosant replied:
It was failing with a 5-second timeout because it was a longer test. initially, I also implemented a frozen phase and tested it here. Since we removed the frozen phase, I also removed a timeout, as I expect it should fit in 5 seconds now.

alisonelizabeth commented on x-pack/plugins/index_lifecycle_management/public/application/sections/edit_policy/components/phases/shared_fields/downsample_field.tsx:
```diff
@@ -0,0 +1,82 @@
+            defaultMessage="Roll up documents within a fixed interval to a single summary document. Reduces the index footprint by storing time series data at reduced granularity."
+          />{' '}
+          {/* TODO: add when available*/}
+          {/* <LearnMoreLink docPath={docLinks.links.elasticsearch.ilmDownsample} /> */}
```
Is this documentation planned, or already in progress? I'm wondering if it makes sense to remove this code until it is ready.

Dosant replied:
I believe this is planned: https://github.com/elastic/kibana/issues/130437#issuecomment-1226598533 cc @debadair

alisonelizabeth commented on x-pack/plugins/index_lifecycle_management/public/application/sections/edit_policy/constants.ts:
```diff
@@ -118,3 +118,42 @@ export const timeUnits = [
+/*
+ * Labels for fixed intervals
+ */
+export const fixedIntervalUnits = [
```
I see we already have a `timeUnits` const. Just want to confirm these values should be separate?

Dosant replied:
micros and nanos are technically not supported for the fixed interval. 
btw, I saw this: https://github.com/elastic/kibana/issues/137081

I think we can unite these consts if we remove everything smaller then a minute

alisonelizabeth commented on x-pack/plugins/index_lifecycle_management/__jest__/client_integration/edit_policy/form_validation/downsample_interval.test.ts:
```diff
@@ -0,0 +1,130 @@
+/*
```
Thanks for adding these tests 👍

alisonelizabeth reviewed (approved):
Nice work @Dosant!

I agree with Yulia and JL's comments on looking into ways we can improve the usability of the form in a future iteration. 

I left some minor code comments, nothing blocking.

> Also is the Downsampling action also available in the Frozen phase?
> 
> No! only hot,warm,cold

I noticed in your [last comment on the issue](https://github.com/elastic/kibana/issues/130437#issuecomment-1233088029), you mentioned it _should_ be available in the frozen phase, but is not yet available in ES. Is this future work?

Dosant commented:
> I noticed in your https://github.com/elastic/kibana/issues/130437#issuecomment-1233088029, you mentioned it should be available in the frozen phase, but is not yet available in ES. Is this future work?

@alisonelizabeth, 

There was some back-and-forth :( 
Initially, we planned hot,warm,cold. 
Then we agreed to add frozen. I added frozen. 
Then we agreed not to add frozen (it turned out to be more difficult on es side than originally thought). I removed the frozen. 
The last state: keep hot,warm,cold

alisonelizabeth commented:
@Dosant I forgot to ask during my review -  I noticed you have the `release_note:skip` label on this. However, this seems like something we might want to add a release note for?
